File: files/CH-109.R
library(tidyverse)
library(readxl)

path = "files/CH-109 Custom Grouping.xlsx"
input = read_excel(path, range = "B2:C26")
test  = read_excel(path, range = "G2:H10")
  

result = input %>%
  mutate(group = cumsum(Sales < lag(Sales, default = first(Sales))),
         Date = format(Date, "%m/%d/%Y")) %>%
  summarise(range = paste0(min(Date),"-",max(Date)),
            `Total Sales` = sum(Sales),
            .by = group) 

identical(result$`Total Sales`, test$`Total Sales`)
#> [1] TRUE

File: files/100-199/109/CH-109.py
import pandas as pd
import numpy as np

path = "CH-109 Custom Grouping.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1)
test  = pd.read_excel(path, usecols="G:H", skiprows=1, nrows = 8)

result = input.assign(group = input['Sales'].lt(input['Sales'].shift().fillna(input['Sales'].iloc[0])).cumsum(),
                      Date = input['Date'].dt.strftime("%m/%d/%Y")).groupby('group').agg(range=('Date', lambda x: f"{x.min()}-{x.max()}"),
                                                                                       Total_Sales=('Sales', 'sum')).reset_index()

print(result["Total_Sales"].equals(test["Total Sales"])) # True